File: 116_PopulatingNextRightPointersInEachNode.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root:
            return
        if not root.right:
            self.connect(root.left)
            return
        self.connect(root.left)
        self.connect(root.right)
        lchild, rchild = root.left, root.right
        while rchild:
            lchild.next = rchild
            lchild, rchild = lchild.right, rchild.left

##40.2%			
class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root:
            return
        pre, curr = root, None
        while pre.left:
            curr = pre
            while curr:
                curr.left.next = curr.right
                if curr.next:
                    curr.right.next = curr.next.left
                curr = curr.next
            pre = pre.left

File: 116_PopulatingNextRightPointersInEachNode.cpp
//40.58%
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if(!root)
			return;
		if(!root -> right){
			connect(ro0t -> left);
			return;
		}
		connect(root -> left);
		connect(root -> right);
		TreeLinkNode *lchild(root -> left), *rchild(root -> right);
		while(rchild){
			lchild -> next = rchild;
			lchild = lchild -> right;
			rchild = rchild -> left;
		}	
    }
};

//An iterative solution from LeetCode
void connect(TreeLinkNode *root) {
    if (root == NULL) return;
    TreeLinkNode *pre = root;
    TreeLinkNode *cur = NULL;
    while(pre->left) {
        cur = pre;
        while(cur) {
            cur->left->next = cur->right;
            if(cur->next) cur->right->next = cur->next->left;
            cur = cur->next;
        }
        pre = pre->left;
    }
}